src/app/components/MainContent.tsx

```diff
@@ -1,17 +1,118 @@
-      ref={registerMainContent}
-      tabIndex={-1}
-    >
-      {children}
-    </div>}
-  </Consumer>;
+interface PropTypes {
+  className?: string;
+}
+
+class MainContent extends Component<PropTypes> {
+
+  private rootEl: Element | null = null;
+
+  public componentDidMount() {
+    if (this.rootEl) { this.addGenericJs(this.rootEl); }
+  }
+
+  public componentDidUpdate() {
+    if (this.rootEl) { this.addGenericJs(this.rootEl); }
+  }
+
+  // from https://github.com/openstax/webview/blob/f95b1d0696a70f0b61d83a85c173102e248354cd
+  // .../src/scripts/modules/media/body/body.coffee#L123
+  public addGenericJs(rootEl: Element) {
+    this.addScopeToTables(rootEl);
+    this.wrapElements(rootEl);
+    this.tweakFigures(rootEl);
+    this.addNoFollow(rootEl);
+    this.fixLists(rootEl);
+  }
+
+  public addScopeToTables(rootEl: Element) {
+    rootEl.querySelectorAll('table th').forEach((el) => el.setAttribute('scope', 'col'));
+  }
+
+  // Wrap title and content elements in header and section elements, respectively
+  public wrapElements(rootEl: Element) {
+  rootEl.querySelectorAll(`.example, .exercise, .note, .abstract,
+    [data-type="example"], [data-type="exercise"],
+    [data-type="note"], [data-type="abstract"]`).forEach((el) => {
+
+      // JSDOM does not support `:scope` in .querySelectorAll() so use .matches()
+      const titles = Array.from(el.children).filter(child => child.matches('.title, [data-type="title"], .os-title'))
+
+      const bodyWrap = assertDefined(document, 'document should be defined').createElement('section');
+      bodyWrap.append(...Array.from(el.childNodes));
+
+      const titleWrap = assertDefined(document, 'document should be defined').createElement('header');
+      titleWrap.append(...Array.from(titles));
+
+      el.append(titleWrap, bodyWrap);
+
+      // Add an attribute for the parents' `data-label`
+      // since CSS does not support `parent(attr(data-label))`.
+      // When the title exists, this attribute is added before it
+      const label = el.getAttribute('data-label');
+      if (label) {
+        titles.forEach((title) => title.setAttribute('data-label-parent', label));
+      }
+
+      // Add a class for styling since CSS does not support `:has(> .title)`
+      // NOTE: `.toggleClass()` explicitly requires a `false` (not falsy) 2nd argument
+      if (titles.length > 0) {
+        el.classList.add('ui-has-child-title');
+      }
+    });
+  }
+
+  public tweakFigures(rootEl: Element) {
+    // move caption to bottom of figure
+    rootEl.querySelectorAll('figure > figcaption').forEach((el) => {
+      const parent = assertDefined(el.parentElement, 'figcaption parent should always be defined');
+      parent.classList.add('ui-has-child-figcaption');
+      parent.appendChild(el)
+    });
+  }
+
+  // Add nofollow to external user-generated links
+  public addNoFollow(rootEl: Element) {
+    rootEl.querySelectorAll('a[href^="http:"], a[href^="https:"], a[href^="//"]')
```

Comment:
i'm thinking maybe this is unnecessarily specific, not sure there is a reason to allow following on any links in the content